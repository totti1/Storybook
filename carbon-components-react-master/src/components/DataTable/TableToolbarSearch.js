import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import Search from '../Search';

const TableToolbarSearch = ({
  className,
  searchContainerClass,
  onChange,
  ...rest
}) => {
  const searchContainerClasses = cx(
    searchContainerClass,
    'bx--toolbar-search-container'
  );
  return (
    <div className={searchContainerClasses}>
      <Search
        className={className}
        {...rest}
        small
        id="search-2"
        labelText="Filter table"
        placeHolderText="Search"
        onChange={onChange}
      />
    </div>
  );
};

TableToolbarSearch.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  /**
   * Provide an optional class name for the search container
   */
  searchContainerClasses: PropTypes.string,

  /**
   * Provide an optional hook that is called each time the input is updated
   */
  onChange: PropTypes.func,
};

export default TableToolbarSearch;
