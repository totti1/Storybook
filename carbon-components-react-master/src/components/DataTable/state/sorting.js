import { sortRows } from '../tools/sorting';

/**
 * We currently support the following sorting states for DataTable headers,
 * namely: `NONE` for no sorting being applied, and then `DESC` and `ASC` for
 * the corresponding direction of the sorting order.
 */
export const sortStates = {
  NONE: 'NONE',
  DESC: 'DESC',
  ASC: 'ASC',
};

// Our initialSortState should be `NONE`, unless a consumer has specified a
// different initialSortState
export const initialSortState = sortStates.NONE;

/**
 * Utility used to get the next sort state given the following pieces of
 * information:
 *
 * @param {string} prevHeader the value of the previous header
 * @param {string} header the value of the currently selected header
 * @param {string} prevState the previous sort state of the table
 * @returns {string}
 */
export const getNextSortDirection = (prevHeader, header, prevState) => {
  // If the previous header is equivalent to the current header, we know that we
  // have to derive the next sort state from the previous sort state
  if (prevHeader === header) {
    // When transitioning, we know that the sequence of states is as follows:
    // NONE -> DESC -> ASC -> NONE
    if (prevState === 'NONE') {
      return sortStates.DESC;
    }
    if (prevState === 'DESC') {
      return sortStates.ASC;
    }
    return sortStates.NONE;
  }
  // Otherwise, we have selected a new header and need to start off by sorting
  // in descending order by default
  return sortStates.DESC;
};

/**
 * Derive the next set of sort state fields from props and state for the given
 * header key.
 *
 * @param {Object} props
 * @param {string} props.locale The current locale
 * @param {Function} props.sortRows Method to handle sorting a collection of
 * rows
 * @param {Object} state
 * @param {Array<string>} state.rowIds Array of row ids
 * @param {Object} state.cellsById Lookup object for cells by id
 * @param {string} state.sortDirection The current sort direction
 * @param {string} state.sortHeaderKey The current sort header ky
 * @param {Array<string>} state.initialRowOrder Initial row order for the
 * current set of rows
 * @param {Object} options
 * @param {string} options.key the key for the given header we are derving the
 * next sort state for
 * @returns {Object}
 */
export const getNextSortState = (props, state, { key }) => {
  const {
    rowIds,
    cellsById,
    sortDirection,
    sortHeaderKey,
    initialRowOrder,
  } = state;
  const { locale, sortRow } = props;
  const nextSortDirection = getNextSortDirection(
    key,
    sortHeaderKey,
    sortDirection
  );
  const nextRowIds =
    nextSortDirection !== sortStates.NONE
      ? sortRows({
          rowIds,
          cellsById,
          sortDirection,
          key,
          locale,
          sortRow,
        })
      : initialRowOrder;
  return {
    sortHeaderKey: key,
    sortDirection: nextSortDirection,
    rowIds: nextRowIds,
  };
};
