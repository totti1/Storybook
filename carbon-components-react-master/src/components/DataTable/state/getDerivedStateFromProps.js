import { initialSortState } from './sorting';
import normalize from '../tools/normalize';

/**
 * Helper to derive the next state from the given props and the
 * prevState. Potential future-facing API hook for React v17.
 *
 * Currently, it's being used as a way to normalize the incoming data that we
 * are receiving for rows
 */
const getDerivedStateFromProps = props => {
  const { rowIds, rowsById, cellsById } = normalize(props.rows, props.headers);
  return {
    rowIds,
    rowsById,
    cellsById,
    sortDirection: initialSortState,
    sortHeaderKey: null,
    // Copy over rowIds so the reference doesn't mutate the stored
    // `initialRowOrder`
    initialRowOrder: rowIds.slice(),

    filterInputValue: null,

    // Optional state field to indicate whether a consumer should show a
    // batch actions menu
    shouldShowBatchActions: false,
  };
};

export default getDerivedStateFromProps;
