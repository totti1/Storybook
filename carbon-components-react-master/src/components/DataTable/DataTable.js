import PropTypes from 'prop-types';
import React from 'react';
import getDerivedStateFromProps from './state/getDerivedStateFromProps';
import { getNextSortState } from './state/sorting';
import denormalize from './tools/denormalize';
import { composeEventHandlers } from './tools/events';
import { defaultFilterRows } from './tools/filter';
import { defaultSortRow } from './tools/sorting';
import setupGetInstanceId from './tools/instanceId';

const getInstanceId = setupGetInstanceId();

const translationKeys = {
  expandRow: 'carbon.table.row.expand',
  collapseRow: 'carbon.table.row.collapse',
  selectAll: 'carbon.table.all.select',
  unselectAll: 'carbon.table.all.unselect',
  selectRow: 'carbon.table.row.select',
  unselectRow: 'carbon.table.row.unselect',
};

const defaultTranslations = {
  [translationKeys.expandRow]: 'Expand current row',
  [translationKeys.collapseRow]: 'Collapse current row',
  [translationKeys.selectAll]: 'Select all rows',
  [translationKeys.unselectAll]: 'Unselect all rows',
  [translationKeys.selectRow]: 'Select row',
  [translationKeys.unselectRow]: 'Unselect row',
};

const translateWithId = id => defaultTranslations[id];

/**
 * Data Tables are used to represent a collection of resources, displaying a
 * subset of their fields in columns, or headers. We prioritize direct updates
 * to the state of what we're rendering, so internally we end up normalizing the
 * given data and then denormalizing it when rendering.
 *
 * As a result, each part of the DataTable is accessible through look-up by id,
 * and updating the state of the single entity will cascade updates to the
 * consumer.
 */
export default class DataTable extends React.Component {
  static propTypes = {
    /**
     * The `rows` prop is where you provide us with a list of all the rows that
     * you want to render in the table. The only hard requirement is that this
     * is an array of objects, and that each object has a unique `id` field
     * available on it.
     */
    rows: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
      })
    ).isRequired,

    /**
     * The `headers` prop represents the order in which the headers should
     * appear in the table. We expect an array of objects to be passed in, where
     * `key` is the name of the key in a row object, and `header` is the name of
     * the header.
     */
    headers: PropTypes.arrayOf(
      PropTypes.shape({
        key: PropTypes.string.isRequired,
        header: PropTypes.string.isRequired,
      })
    ).isRequired,

    /**
     * Optional hook to manually control sorting of the rows.
     */
    sortRow: PropTypes.func,

    /**
     * Optional hook to manually control filtering of the rows from the
     * TableToolbarSearch component
     */
    filterRows: PropTypes.func,

    /**
     * Provide a string for the current locale
     */
    locale: PropTypes.string,

    /**
     * Optional method that takes in a message id and returns an
     * internationalized string. See `DataTable.translationKeys` for all
     * available message ids.
     */
    translateWithId: PropTypes.func,
  };

  static defaultProps = {
    sortRow: defaultSortRow,
    filterRows: defaultFilterRows,
    locale: 'en',
    translateWithId,
  };

  static translationKeys = Object.values(translationKeys);

  constructor(props) {
    super(props);
    this.state = getDerivedStateFromProps(props);
    this.instanceId = getInstanceId();
  }

  componentWillReceiveProps(nextProps) {
    const nextState = getDerivedStateFromProps(nextProps);
    if (nextState) {
      this.setState(nextState);
    }
  }

  /**
   * Get the props associated with the given header. Mostly used for adding in
   * sorting behavior.
   *
   * @param {Object} config
   * @param {string} config.header the header we want the props for
   * @param {Function} config.onClick a custom click handler for the header
   * @returns {Object}
   */
  getHeaderProps = ({ header, onClick, ...rest }) => {
    const { sortDirection, sortHeaderKey } = this.state;
    return {
      ...rest,
      key: header.key,
      sortDirection,
      isSortable: true,
      isSortHeader: sortHeaderKey === header.key,
      // Compose the event handlers so we don't overwrite a consumer's `onClick`
      // handler
      onClick: composeEventHandlers([this.handleSortBy(header.key), onClick]),
    };
  };

  /**
   * Get the props associated with the given row. Mostly used for expansion.
   *
   * @param {Object} config
   * @param {Object} config.row the row we want the props for
   * @param {Function} config.onClick a custom click handler for the header
   * @returns {Object}
   */
  getRowProps = ({ row, onClick, ...rest }) => {
    const { translateWithId: t } = this.props;
    const translationKey = row.isExpanded
      ? translationKeys.collapseRow
      : translationKeys.expandRow;
    return {
      ...rest,
      key: row.id,
      // Compose the event handlers so we don't overwrite a consumer's `onClick`
      // handler
      onExpand: composeEventHandlers([this.handleOnExpandRow(row.id), onClick]),
      isExpanded: row.isExpanded,
      ariaLabel: t(translationKey),
    };
  };

  /**
   * Gets the props associated with selection for a header or a row, where
   * applicable. Most often used to indicate selection status of the table or
   * for a specific row.
   *
   * @param {Object} [row] an optional row that we want to access the props for
   * @returns {Object}
   */
  getSelectionProps = ({ onClick, row, ...rest } = {}) => {
    const { translateWithId: t } = this.props;

    // If we're given a row, return the selection state values for that row
    if (row) {
      const translationKey = row.isSelected
        ? translationKeys.unselectRow
        : translationKeys.selectRow;
      return {
        ...rest,
        checked: row.isSelected,
        onSelect: composeEventHandlers([
          this.handleOnSelectRow(row.id),
          onClick,
        ]),
        id: `${this.getTablePrefix()}__select-row-${row.id}`,
        name: `select-row-${row.id}`,
        ariaLabel: t(translationKey),
      };
    }

    // Otherwise, we're working on `TableSelectAll` which handles toggling the
    // selection state of all rows
    const checked = this.getSelectedRows().length === this.state.rowIds.length;
    const translationKey = checked
      ? translationKeys.unselectAll
      : translationKeys.selectAll;
    return {
      ...rest,
      checked,
      onSelect: composeEventHandlers([this.handleSelectAll, onClick]),
      id: `${this.getTablePrefix()}__select-all`,
      name: 'select-all',
      ariaLabel: t(translationKey),
    };
  };

  getBatchActionProps = (props = {}) => {
    const { shouldShowBatchActions } = this.state;
    const totalSelected = this.getSelectedRows().length;
    return {
      ...props,
      shouldShowBatchActions,
      totalSelected,
      onCancel: this.handleOnCancel,
    };
  };

  /**
   * Helper utility to get all the currently selected rows
   * @returns {Array<string>} the array of rowIds that are currently selected
   */
  getSelectedRows = () =>
    this.state.rowIds.filter(id => {
      const row = this.state.rowsById[id];
      return row.isSelected;
    });

  /**
   * Helper for getting the table prefix for elements that require an
   * `id` attribute that is unique.
   *
   * @returns {string}
   */
  getTablePrefix = () => `data-table-${this.instanceId}`;

  /**
   * Handler for the `onCancel` event to hide the batch action bar
   */
  handleOnCancel = () => {
    this.setState({ shouldShowBatchActions: false });
  };

  /**
   * Handler for toggling the selection state of all rows in the database
   */
  handleSelectAll = () => {
    this.setState(state => {
      const { rowIds } = state;
      const isSelected = this.getSelectedRows().length !== rowIds.length;
      return {
        shouldShowBatchActions: isSelected,
        rowsById: rowIds.reduce(
          (acc, id) => ({
            ...acc,
            [id]: {
              ...state.rowsById[id],
              isSelected,
            },
          }),
          {}
        ),
      };
    });
  };

  /**
   * Handler for toggling the selection state of a given row.
   *
   * @param {string} rowId
   * @returns {Function}
   */
  handleOnSelectRow = rowId => () => {
    this.setState(state => {
      const row = state.rowsById[rowId];
      const selectedRows = state.rowIds.filter(id => {
        return state.rowsById[id].isSelected;
      }).length;
      // Predict the length of the selected rows after this change occurs
      const selectedRowsCount = !row.isSelected
        ? selectedRows + 1
        : selectedRows - 1;
      return {
        // Basic assumption here is that we want to show the batch action bar if
        // the row is being selected. If it's being unselected, then see if we
        // have a non-zero number of selected rows that batch actions could
        // still apply to
        shouldShowBatchActions: !row.isSelected || selectedRowsCount > 0,
        rowsById: {
          ...state.rowsById,
          [rowId]: {
            ...row,
            isSelected: !row.isSelected,
          },
        },
      };
    });
  };

  /**
   * Handler for toggling the expansion state of a given row.
   *
   * @param {string} rowId
   * @returns {Function}
   */
  handleOnExpandRow = rowId => () => {
    this.setState(state => {
      const row = state.rowsById[rowId];
      return {
        rowsById: {
          ...state.rowsById,
          [rowId]: {
            ...row,
            isExpanded: !row.isExpanded,
          },
        },
      };
    });
  };

  /**
   * Handler for transitioning to the next sort state of the table
   *
   * @param {string} headerKey the field for the header that we are sorting by
   * @returns {Function}
   */
  handleSortBy = headerKey => () => {
    this.setState(state =>
      getNextSortState(this.props, state, { key: headerKey })
    );
  };

  /**
   * Event handler for transitioning input value state changes for the table
   * filter component.
   *
   * @param {Event} event
   */
  handleOnInputValueChange = event => {
    this.setState({ filterInputValue: event.target.value });
  };

  render() {
    const { children, filterRows, headers, render } = this.props;
    const { filterInputValue, rowIds, rowsById, cellsById } = this.state;
    const filteredRowIds =
      typeof filterInputValue === 'string'
        ? filterRows({
            rowIds,
            headers,
            cellsById,
            inputValue: filterInputValue,
          })
        : rowIds;
    const renderProps = {
      // Data derived from state
      rows: denormalize(filteredRowIds, rowsById, cellsById),
      headers: this.props.headers,
      selectedRows: denormalize(this.getSelectedRows(), rowsById, cellsById),

      // Prop accessors/getters
      getHeaderProps: this.getHeaderProps,
      getRowProps: this.getRowProps,
      getSelectionProps: this.getSelectionProps,
      getBatchActionProps: this.getBatchActionProps,

      // Custom event handlers
      onInputChange: this.handleOnInputValueChange,

      // Expose internal state change actions
      sortBy: headerKey => this.handleSortBy(headerKey)(),
      selectAll: this.handleSelectAll,
      selectRow: rowId => this.handleOnSelectRow(rowId)(),
      expandRow: rowId => this.handleOnExpandRow(rowId)(),
    };

    if (render !== undefined) {
      return render(renderProps);
    }

    if (children !== undefined) {
      return children(renderProps);
    }

    return null;
  }
}
