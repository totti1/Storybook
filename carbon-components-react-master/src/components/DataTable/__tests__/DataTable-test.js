import React from 'react';
import Button from '../../Button';
import DataTable, {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableHeader,
  TableRow,
  TableToolbar,
  TableToolbarAction,
  TableToolbarContent,
  TableToolbarSearch,
} from '../';
import { sortStates } from '../state/sorting';
import { mount } from 'enzyme';

const getHeaderAt = (wrapper, index) =>
  wrapper.find('TableHeader button').at(index);
const getFilterInput = wrapper => wrapper.find('TableToolbarSearch input');

describe('DataTable', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      rows: [
        {
          id: 'b',
          fieldA: 'Field 2:A',
          fieldB: 'Field 2:B',
        },
        {
          id: 'a',
          fieldA: 'Field 1:A',
          fieldB: 'Field 1:B',
        },
        {
          id: 'c',
          fieldA: 'Field 3:A',
          fieldB: 'Field 3:B',
        },
      ],
      headers: [
        {
          key: 'fieldA',
          header: 'Field A',
        },
        {
          key: 'fieldB',
          header: 'Field B',
        },
      ],
      locale: 'en',
      render: jest.fn(({ rows, headers, getHeaderProps, onInputChange }) => (
        <TableContainer title="DataTable with toolbar">
          <TableToolbar>
            <TableToolbarSearch onChange={onInputChange} />
            <TableToolbarContent>
              <TableToolbarAction
                iconName="download"
                iconDescription="Download"
                onClick={jest.fn()}
              />
              <TableToolbarAction
                iconName="edit"
                iconDescription="Edit"
                onClick={jest.fn()}
              />
              <TableToolbarAction
                iconName="settings"
                iconDescription="Settings"
                onClick={jest.fn()}
              />
              <Button onClick={jest.fn()} small kind="primary">
                Add new
              </Button>
            </TableToolbarContent>
          </TableToolbar>
          <Table>
            <TableHead>
              <TableRow>
                {headers.map(header => (
                  <TableHeader {...getHeaderProps({ header })}>
                    {header.header}
                  </TableHeader>
                ))}
              </TableRow>
            </TableHead>
            <TableBody>
              {rows.map(row => (
                <TableRow key={row.id}>
                  {row.cells.map(cell => (
                    <TableCell key={cell.id}>{cell.value}</TableCell>
                  ))}
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      )),
    };
  });

  it('should render', () => {
    const wrapper = mount(<DataTable {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('sorting', () => {
    it('should sort a row by a header when a header is clicked', () => {
      const wrapper = mount(<DataTable {...mockProps} />);
      const header = getHeaderAt(wrapper, 0);

      header.simulate('click');
      expect(wrapper.state('rowIds')).toEqual(['a', 'b', 'c']);

      header.simulate('click');
      expect(wrapper.state('rowIds')).toEqual(['c', 'b', 'a']);

      header.simulate('click');
      expect(wrapper.state('rowIds')).toEqual(['b', 'a', 'c']);
    });

    it('should reset to DESC ordering when another header is clicked', () => {
      const wrapper = mount(<DataTable {...mockProps} />);
      const firstHeader = getHeaderAt(wrapper, 0);
      const secondHeader = getHeaderAt(wrapper, 1);

      firstHeader.simulate('click');
      expect(wrapper.state('rowIds')).toEqual(['a', 'b', 'c']);

      firstHeader.simulate('click');
      expect(wrapper.state('rowIds')).toEqual(['c', 'b', 'a']);
      expect(wrapper.state('sortDirection')).toBe(sortStates.ASC);

      secondHeader.simulate('click');
      expect(wrapper.state('sortDirection')).toBe(sortStates.DESC);
    });
  });

  describe('filtering', () => {
    it('should filter rows by the given input', () => {
      const wrapper = mount(<DataTable {...mockProps} />);
      const filterInput = getFilterInput(wrapper);

      expect(wrapper.state('rowIds').length).toBe(mockProps.rows.length);

      filterInput.simulate('change', {
        target: {
          value: 'Field 1',
        },
      });

      expect(mockProps.render).toHaveBeenCalledWith(
        expect.objectContaining({
          rows: [
            expect.objectContaining({
              id: 'a',
            }),
          ],
        })
      );
    });
  });
});
