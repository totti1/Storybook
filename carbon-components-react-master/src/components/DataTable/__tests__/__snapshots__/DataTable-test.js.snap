// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render 1`] = `
<DataTable
  filterRows={[Function]}
  headers={
    Array [
      Object {
        "header": "Field A",
        "key": "fieldA",
      },
      Object {
        "header": "Field B",
        "key": "fieldB",
      },
    ]
  }
  locale="en"
  render={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "expandRow": [Function],
            "getBatchActionProps": [Function],
            "getHeaderProps": [Function],
            "getRowProps": [Function],
            "getSelectionProps": [Function],
            "headers": Array [
              Object {
                "header": "Field A",
                "key": "fieldA",
              },
              Object {
                "header": "Field B",
                "key": "fieldB",
              },
            ],
            "onInputChange": [Function],
            "rows": Array [
              Object {
                "cells": Array [
                  Object {
                    "errors": null,
                    "id": "b:fieldA",
                    "isEditable": false,
                    "isEditing": false,
                    "isValid": true,
                    "value": "Field 2:A",
                  },
                  Object {
                    "errors": null,
                    "id": "b:fieldB",
                    "isEditable": false,
                    "isEditing": false,
                    "isValid": true,
                    "value": "Field 2:B",
                  },
                ],
                "id": "b",
                "isExpanded": false,
                "isSelected": false,
              },
              Object {
                "cells": Array [
                  Object {
                    "errors": null,
                    "id": "a:fieldA",
                    "isEditable": false,
                    "isEditing": false,
                    "isValid": true,
                    "value": "Field 1:A",
                  },
                  Object {
                    "errors": null,
                    "id": "a:fieldB",
                    "isEditable": false,
                    "isEditing": false,
                    "isValid": true,
                    "value": "Field 1:B",
                  },
                ],
                "id": "a",
                "isExpanded": false,
                "isSelected": false,
              },
              Object {
                "cells": Array [
                  Object {
                    "errors": null,
                    "id": "c:fieldA",
                    "isEditable": false,
                    "isEditing": false,
                    "isValid": true,
                    "value": "Field 3:A",
                  },
                  Object {
                    "errors": null,
                    "id": "c:fieldB",
                    "isEditable": false,
                    "isEditing": false,
                    "isValid": true,
                    "value": "Field 3:B",
                  },
                ],
                "id": "c",
                "isExpanded": false,
                "isSelected": false,
              },
            ],
            "selectAll": [Function],
            "selectRow": [Function],
            "selectedRows": Array [],
            "sortBy": [Function],
          },
        ],
      ],
    }
  }
  rows={
    Array [
      Object {
        "fieldA": "Field 2:A",
        "fieldB": "Field 2:B",
        "id": "b",
      },
      Object {
        "fieldA": "Field 1:A",
        "fieldB": "Field 1:B",
        "id": "a",
      },
      Object {
        "fieldA": "Field 3:A",
        "fieldB": "Field 3:B",
        "id": "c",
      },
    ]
  }
  sortRow={[Function]}
  translateWithId={[Function]}
>
  <TableContainer
    title="DataTable with toolbar"
  >
    <div
      className="bx--data-table-v2-container"
    >
      <h4
        className="bx--data-table-v2-header"
      >
        DataTable with toolbar
      </h4>
      <TableToolbar>
        <section
          className="bx--table-toolbar"
        >
          <TableToolbarSearch
            onChange={[Function]}
          >
            <div
              className="bx--toolbar-search-container"
            >
              <Search
                id="search-2"
                labelText="Filter table"
                onChange={[Function]}
                placeHolderText="Search"
                small={true}
                type="text"
              >
                <div
                  className="bx--search bx--search-with-options bx--search--sm"
                  role="search"
                >
                  <Icon
                    className="bx--search-magnifier"
                    description="search"
                    fillRule="evenodd"
                    name="search--glyph"
                    role="img"
                  >
                    <svg
                      alt="search"
                      aria-label="search"
                      className="bx--search-magnifier"
                      fillRule="evenodd"
                      height="16"
                      name="search--glyph"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <title>
                        search
                      </title>
                      <path
                        d="M6 2c2.2 0 4 1.8 4 4s-1.8 4-4 4-4-1.8-4-4 1.8-4 4-4zm0-2C2.7 0 0 2.7 0 6s2.7 6 6 6 6-2.7 6-6-2.7-6-6-6zm10 13.8L13.8 16l-3.6-3.6 2.2-2.2z"
                        key="key0"
                      />
                      <path
                        d="M16 13.8L13.8 16l-3.6-3.6 2.2-2.2z"
                        key="key1"
                      />
                    </svg>
                  </Icon>
                  <label
                    className="bx--label"
                    htmlFor="search-2"
                  >
                    Filter table
                  </label>
                  <input
                    className="bx--search-input"
                    id="search-2"
                    onChange={[Function]}
                    placeholder="Search"
                    type="text"
                  />
                  <Icon
                    className="bx--search-close bx--search-close--hidden"
                    description="close"
                    fillRule="evenodd"
                    name="close--glyph"
                    onClick={[Function]}
                    role="img"
                  >
                    <svg
                      alt="close"
                      aria-label="close"
                      className="bx--search-close bx--search-close--hidden"
                      fillRule="evenodd"
                      height="16"
                      name="close--glyph"
                      onClick={[Function]}
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <title>
                        close
                      </title>
                      <path
                        d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm3.5 10.1l-1.4 1.4L8 9.4l-2.1 2.1-1.4-1.4L6.6 8 4.5 5.9l1.4-1.4L8 6.6l2.1-2.1 1.4 1.4L9.4 8l2.1 2.1z"
                        key="key0"
                      />
                    </svg>
                  </Icon>
                </div>
              </Search>
            </div>
          </TableToolbarSearch>
          <TableToolbarContent>
            <div
              className="bx--toolbar-content"
            >
              <TableToolbarAction
                iconDescription="Download"
                iconName="download"
                onClick={[MockFunction]}
              >
                <button
                  className="bx--toolbar-action"
                  onClick={[MockFunction]}
                >
                  <Icon
                    className="bx--toolbar-action__icon"
                    description="Download"
                    fillRule="evenodd"
                    name="download"
                    role="img"
                  >
                    <svg
                      alt="Download"
                      aria-label="Download"
                      className="bx--toolbar-action__icon"
                      fillRule="evenodd"
                      height="24"
                      name="download"
                      role="img"
                      viewBox="0 0 24 24"
                      width="24"
                    >
                      <title>
                        Download
                      </title>
                      <path
                        d="M19 9.4l-1.2-1.1L13 13V0h-2v13L6.2 8.3 5 9.4l7 6.6z"
                        key="key0"
                      />
                      <path
                        d="M22 14v6H2v-6H0v10h24V14z"
                        key="key1"
                      />
                    </svg>
                  </Icon>
                </button>
              </TableToolbarAction>
              <TableToolbarAction
                iconDescription="Edit"
                iconName="edit"
                onClick={[MockFunction]}
              >
                <button
                  className="bx--toolbar-action"
                  onClick={[MockFunction]}
                >
                  <Icon
                    className="bx--toolbar-action__icon"
                    description="Edit"
                    fillRule="evenodd"
                    name="edit"
                    role="img"
                  >
                    <svg
                      alt="Edit"
                      aria-label="Edit"
                      className="bx--toolbar-action__icon"
                      fillRule="evenodd"
                      height="29"
                      name="edit"
                      role="img"
                      viewBox="0 0 29 29"
                      width="29"
                    >
                      <title>
                        Edit
                      </title>
                      <path
                        d="M20.297 2.932l2.9-2.9 5.806 5.805-2.9 2.9zM3.867 19.385l5.8 5.807 14.5-14.517-5.8-5.806-14.5 14.515zm5.8 3.07L6.6 19.384 18.367 7.607l3.066 3.068-11.766 11.78zM0 29.062l4.833-1.936-2.9-2.903z"
                        key="key0"
                      />
                    </svg>
                  </Icon>
                </button>
              </TableToolbarAction>
              <TableToolbarAction
                iconDescription="Settings"
                iconName="settings"
                onClick={[MockFunction]}
              >
                <button
                  className="bx--toolbar-action"
                  onClick={[MockFunction]}
                >
                  <Icon
                    className="bx--toolbar-action__icon"
                    description="Settings"
                    fillRule="evenodd"
                    name="settings"
                    role="img"
                  >
                    <svg
                      alt="Settings"
                      aria-label="Settings"
                      className="bx--toolbar-action__icon"
                      fillRule="evenodd"
                      height="24"
                      name="settings"
                      role="img"
                      viewBox="0 0 24 24"
                      width="24"
                    >
                      <title>
                        Settings
                      </title>
                      <path
                        d="M21.2 12c0-.5 0-1-.1-1.4L24 8l-2.5-4.1L18 5.1c-.8-.7-1.8-1.3-2.9-1.7L14.4 0H9.6l-.8 3.4c-1.1.4-2 1-2.9 1.7L2.5 3.9 0 8l2.9 2.6c-.1.4-.1.9-.1 1.4 0 .5 0 1 .1 1.4L0 16l2.5 4.1L6 18.9c.8.7 1.8 1.3 2.9 1.7l.7 3.4h4.8l.8-3.4c1.1-.4 2-1 2.9-1.7l3.5 1.1 2.4-4-2.9-2.6c.1-.4.1-.9.1-1.4zM12 15c-1.7 0-3-1.3-3-3s1.3-3 3-3 3 1.3 3 3-1.3 3-3 3z"
                        key="key0"
                      />
                    </svg>
                  </Icon>
                </button>
              </TableToolbarAction>
              <Button
                disabled={false}
                iconDescription="Provide icon description if icon is used"
                kind="primary"
                onClick={[MockFunction]}
                small={true}
                tabIndex={0}
                type="button"
              >
                <button
                  className="bx--btn bx--btn--sm bx--btn--primary"
                  disabled={false}
                  onClick={[MockFunction]}
                  tabIndex={0}
                  type="button"
                >
                  Add new
                </button>
              </Button>
            </div>
          </TableToolbarContent>
        </section>
      </TableToolbar>
      <Table>
        <table
          className="bx--data-table-v2 bx--data-table-v2--zebra"
        >
          <TableHead>
            <thead>
              <TableRow>
                <tr>
                  <TableHeader
                    isSortHeader={false}
                    isSortable={true}
                    key="fieldA"
                    onClick={[Function]}
                    scope="col"
                    sortDirection="NONE"
                    translateWithId={[Function]}
                  >
                    <th
                      scope="col"
                    >
                      <button
                        className="bx--table-sort-v2"
                        onClick={[Function]}
                      >
                        <span
                          className="bx--table-header-label"
                        >
                          Field A
                        </span>
                        <Icon
                          className="bx--table-sort-v2__icon"
                          description="Sort rows by this header in descending order"
                          fillRule="evenodd"
                          name="caret--down"
                          role="img"
                        >
                          <svg
                            alt="Sort rows by this header in descending order"
                            aria-label="Sort rows by this header in descending order"
                            className="bx--table-sort-v2__icon"
                            fillRule="evenodd"
                            height="5"
                            name="caret--down"
                            role="img"
                            viewBox="0 0 10 5"
                            width="10"
                          >
                            <title>
                              Sort rows by this header in descending order
                            </title>
                            <path
                              d="M10 0L5 5 0 0z"
                              key="key0"
                            />
                          </svg>
                        </Icon>
                      </button>
                    </th>
                  </TableHeader>
                  <TableHeader
                    isSortHeader={false}
                    isSortable={true}
                    key="fieldB"
                    onClick={[Function]}
                    scope="col"
                    sortDirection="NONE"
                    translateWithId={[Function]}
                  >
                    <th
                      scope="col"
                    >
                      <button
                        className="bx--table-sort-v2"
                        onClick={[Function]}
                      >
                        <span
                          className="bx--table-header-label"
                        >
                          Field B
                        </span>
                        <Icon
                          className="bx--table-sort-v2__icon"
                          description="Sort rows by this header in descending order"
                          fillRule="evenodd"
                          name="caret--down"
                          role="img"
                        >
                          <svg
                            alt="Sort rows by this header in descending order"
                            aria-label="Sort rows by this header in descending order"
                            className="bx--table-sort-v2__icon"
                            fillRule="evenodd"
                            height="5"
                            name="caret--down"
                            role="img"
                            viewBox="0 0 10 5"
                            width="10"
                          >
                            <title>
                              Sort rows by this header in descending order
                            </title>
                            <path
                              d="M10 0L5 5 0 0z"
                              key="key0"
                            />
                          </svg>
                        </Icon>
                      </button>
                    </th>
                  </TableHeader>
                </tr>
              </TableRow>
            </thead>
          </TableHead>
          <TableBody>
            <tbody>
              <TableRow
                key="b"
              >
                <tr>
                  <TableCell
                    key="b:fieldA"
                  >
                    <td>
                      Field 2:A
                    </td>
                  </TableCell>
                  <TableCell
                    key="b:fieldB"
                  >
                    <td>
                      Field 2:B
                    </td>
                  </TableCell>
                </tr>
              </TableRow>
              <TableRow
                key="a"
              >
                <tr>
                  <TableCell
                    key="a:fieldA"
                  >
                    <td>
                      Field 1:A
                    </td>
                  </TableCell>
                  <TableCell
                    key="a:fieldB"
                  >
                    <td>
                      Field 1:B
                    </td>
                  </TableCell>
                </tr>
              </TableRow>
              <TableRow
                key="c"
              >
                <tr>
                  <TableCell
                    key="c:fieldA"
                  >
                    <td>
                      Field 3:A
                    </td>
                  </TableCell>
                  <TableCell
                    key="c:fieldB"
                  >
                    <td>
                      Field 3:B
                    </td>
                  </TableCell>
                </tr>
              </TableRow>
            </tbody>
          </TableBody>
        </table>
      </Table>
    </div>
  </TableContainer>
</DataTable>
`;
